Finished the first working version of the API.

radiantdinosaurs opened the pull request:
Created functions that check base64 file type (making sure it is JPEG or PNG), decode the file, and save to disk. Replaced basic image recognition (default image in classify_image.py) for image recognition of user supplied base64 encoded image. Added in proper error responses for unsupported file types and undefined body.

radiantdinosaurs committed:
Finished the first working version of the API.

Created functions that check base64 file type (making sure it is JPEG or PNG), decode the file, and save to disk. Replaced basic image recognition (default image in classify_image.py) for image recognition of user supplied base64 encoded image. Added in proper error responses for unsupported file types and undefined body.

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+function writeJpegToDisk(base64String) {
+    let base64Data = base64String.replace(/^data:image\/jpeg;base64,/, ""); // stripping for decoding
+    // writing file to disk as "out.jpeg"
+    fs.writeFile("out.jpeg", base64Data, "base64", function (error) {
```
This wouldn't handle parallel server requests. If two requests came in, it's possible you're operating on the wrong "out.jpeg". I recommend using the JS equivalent of Java's `UUID.randomUUID()`

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+    fs.writeFile("out.jpeg", base64Data, "base64", function (error) {
+        if (error !== null) {
+            console.log(error);
+        }
```
You should probably handle the error somehow. Just letting it silently fail is asking for trouble

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+        if (error !== null) {
+            console.log(error);
+        }
+    });
```
Have you considered using the arrow syntax instead of an anonymous function?

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+        fs.writeFile("out.jpeg", output, function (error) {
+            if (error !== null) {
+                console.log(error);
+            }
```
Same point here about the silent failure

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+    // using middleware to change image from PNG to JPEG
+    pngToJpeg({quality: 90})(buffer).then(function (output) {
+        // writing file to disk as "out.jpeg"
+        fs.writeFile("out.jpeg", output, function (error) {
```
Same comment about unique filenames

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+    } else if (fileType.includes("jpeg")) {
+        return "JPEG"; // image is a JPEG
+    } else {
+        return "undefined"; // image is not PNG or JPEG, so set as "undefined," because it cannot be used
```
I think you can just return `undefined` instead of the string, `"undefined"`

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+ */
+function checkFileType(base64String) {
+    let fileType = base64String.substring(0, 15); // getting the data/image type label
+    if (fileType.includes("png")) {
```
This throws an exception if fileType is undefined. This function should be self-contained for the most part. If the logic is copied anywhere, someone won't think that this logic isn't enforcing non-null parameters

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
-    let resultLines = result.toString().split("\n");
-    let resultObject = {"image_labels": []};
+    let sloth = false; // true if image contains a sloth
+    let resultLines = result.toString().split("\n"); // splitting up TensorFlow's result by newline
```
Consider adding handling for where result is undefined

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+                if(fields[1].replace(/[^\d.]/g, "") > 0.70) {
+                    sloth = true;
+                }
+            }
```
You should definitely have unit tests for this logic. There may be edge cases you're not handling

andrewvora replied:
Consider using arrow syntax here too for the callback passed to `forEach`

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+        return null; // return null if the file type isn't PNG or JPEG
+    }
+    return "../slothbucket/out.jpeg"; // return file path if the file type is PNG or JPEG
+}
```
Same point here about `fileType` being `undefined`

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+    } else {
+        return null; // return null if the file type isn't PNG or JPEG
+    }
+    return "../slothbucket/out.jpeg"; // return file path if the file type is PNG or JPEG
```
Potential danger with `out.jpeg` being non-unique

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
-        res.send(result.toString().split("\n"));
-    });
+// HTTP POST for any TensorFlow image recognition with bonus sloth check
+app.post("/sloth-check", function (req, res) {
```
Consider using arrow-syntax

andrewvora commented on app.js:
```diff
@@ -1,46 +1,138 @@
+    else {
+        // body was undefined or null
+        res.status(500).send({status:500, message:"Cannot read undefined body"});
+    }
```
This logic as well as the associated functions might be better fit being put into a module that gets required in `app.js`

andrewvora reviewed (changes requested):
Left some comments regarding code style and structure

andrewvora reviewed (changes requested):
You may want to consider putting the app.js file into an `app/` folder. Then update `package.json` to have `start` run `node app/app.js`

radiantdinosaurs committed:
Finished working on request changes.

Added in error handling and some arrow functions. Moved app.js to its own folder and updated the package.json.

andrewvora reviewed (approved):
The logic is a mess and it's hard to read, but given that you're refactoring this into "modules" that can be addressed at a later date. I'm approving this based on the fact the happy path works

Some quick TODOs for a future PR:

* break error handling into their own functions or controllers
* avoid long inline conditionals, i.e. if ( A && B && C && D && E). Prefer `const isF = A && B && C && D && E`.
* Your null and undefined checks don't work the way you think they do

andrewvora commented:
Follow these steps. I leave finding out the commands up to you. Make sure you learn about unfamiliar commands and understand each step. Don't blindly follow steps.

1. Make sure the development branch is up-to-date.
2. Check out your feature branch.
3. Rebase the development branch ONTO your feature branch.
4. Take care of any merge conflicts.
5. Finish the rebase.
6. Push up your changes. You will need to force a push since you will be altering the git history.

Alternatively, you can just do steps 1-2, then merge development ONTO your feature branch. 

Read about the differences between merging and rebasing: https://www.atlassian.com/git/tutorials/merging-vs-rebasing

I prefer rebasing.
